lib/sycamore/sycamore/writers/opensearch.py

```diff
@@ -1,77 +1,184 @@
+        assert isinstance(
+            params, OpenSearchClientParams
+        ), f"Provided params was not of type OpenSearchClientParams:\n{params}"
+        paramsdict = asdict(params)
+        allowed_failures_per_block = paramsdict.pop("number_of_allowed_failures_per_block")
+        failures_fp = paramsdict.pop("collect_failures_filepath")
+        os_client = OpenSearch(**paramsdict)
+        os_client.ping()
+        return OpenSearchClient(os_client, allowed_failures_per_block, failures_fp)
+
+    def write_many_records(self, records: list[BaseDBWriter.Record], target_params: BaseDBWriter.TargetParams):
+        assert isinstance(
+            target_params, OpenSearchTargetParams
+        ), f"Provided target_params was not of type OpenSearchTargetParams:\n{target_params}"
+        assert _narrow_list_of_os_records(records), f"A provided record was not of type OpenSearchRecord:\n{records}"
+        failures = []
+        for success, info in parallel_bulk(self._client, [asdict(r) for r in records]):
+            if not success:
+                log.error("A Document failed to upload", info)
+                failures.append(info)
+
+        if len(failures) > self._allowed_failures_per_block:
+            with open(self._failure_filepath, "a") as f:
+                for doc in failures:
+                    f.write(f"{doc}\n")
+            raise RuntimeError(
+                f"{self._allowed_failures_per_block} documents failed to index. " f"Refer to {self._failure_filepath}."
+            )
+
+    def create_target_idempotent(self, target_params: BaseDBWriter.TargetParams):
+        assert isinstance(
+            target_params, OpenSearchTargetParams
+        ), f"Provided target_params was not of type OpenSearchTargetParams:\n{target_params}"
+        index_name = target_params.index_name
+        try:
+            self._client.indices.create(
+                index_name, body={"mappings": target_params.mappings, "settings": target_params.settings}
+            )
+        except RequestError as e:
+            if e.error != "resource_already_exists_exception":
+                raise e
+
+    def get_existing_target_params(self, target_params: BaseDBWriter.TargetParams) -> OpenSearchTargetParams:
+        def _string_values_to_python_types(obj: Any):
+            if isinstance(obj, dict):
+                for k in obj:
+                    obj[k] = _string_values_to_python_types(obj[k])
+                return obj
+            if isinstance(obj, list):
+                for i in range(len(obj)):
+                    obj[i] = _string_values_to_python_types(obj[i])
+                return obj
+            if isinstance(obj, str):
+                if obj == "true":
+                    return True
+                elif obj == "false":
+                    return False
+                elif obj.isnumeric():
+                    return int(obj)
+                elif re.match("[0-9]+\.[0-9]+", obj):
```

Comment:
apparently json doesn't allow things like .3 and since opensearch uses a standard json lib for all that I'm gonna say I get the 0. That said I dropped all the negative numbers